File: core/ir/semantic_runtime_ir.py
from __future__ import annotations

from typing import Any, Dict


def compile_semantic_runtime_ir(
    cognition: Dict[str, Any],
) -> Dict[str, Any]:
    return {
        "ir": "semantic_runtime",
        "ontology": cognition.get("ontology", {}),
        "entities": cognition.get("entities", {}),
        "semantic_graph": cognition.get("semantic_graph", {}),
        "domain": cognition.get("domain", {}),
        "ui_semantics": cognition.get("ui", {}),
        "table_semantics": cognition.get("tables", {}),
        "document_semantics": cognition.get("document", {}),
        "repository_semantics": cognition.get("repository", {}),
        "application_semantics": cognition.get("application", {}),
        "causality_semantics": cognition.get("causality", {}),
        "workflow_semantics": cognition.get("workflow", {}),
        "alignment": cognition.get("alignment", {}),
        "bounded": True,
    }


def semantic_runtime_ir_to_graph(
    semantic_ir: Dict[str, Any],
) -> Dict[str, Any]:
    graph = semantic_ir.get("semantic_graph", {})
    nodes = list(graph.get("nodes", []))
    edges = list(graph.get("edges", []))

    ontology = semantic_ir.get("ontology", {})
    if ontology.get("primary_domain"):
        nodes.append({
            "id": f"domain:{ontology['primary_domain']}",
            "type": "domain",
        })

    if not nodes:
        nodes = [{"id": "semantic:root", "type": "semantic"}]

    return {
        "ir": "semantic_runtime_graph",
        "nodes": sorted(nodes, key=lambda item: str(item.get("id", ""))),
        "edges": edges,
        "bounded": True,
    }

File: src/ir/semanticRuntimeIr.ts
/**
 * Converted from Python: core/ir/semantic_runtime_ir.py
 * @generated — WebWeaveX python→javascript library port
 */

import * as py from "../runtime/pyCompat.js";

export function compileSemanticRuntimeIr(cognition: any): any {
  return {"ir": "semantic_runtime", "ontology": py.get(cognition, "ontology", {}), "entities": py.get(cognition, "entities", {}), "semantic_graph": py.get(cognition, "semantic_graph", {}), "domain": py.get(cognition, "domain", {}), "ui_semantics": py.get(cognition, "ui", {}), "table_semantics": py.get(cognition, "tables", {}), "document_semantics": py.get(cognition, "document", {}), "repository_semantics": py.get(cognition, "repository", {}), "application_semantics": py.get(cognition, "application", {}), "causality_semantics": py.get(cognition, "causality", {}), "workflow_semantics": py.get(cognition, "workflow", {}), "alignment": py.get(cognition, "alignment", {}), "bounded": true};
}
export function semanticRuntimeIrToGraph(semantic_ir: any): any {
  var graph: any = py.get(semantic_ir, "semantic_graph", {});
  var nodes: any = [...py.iter(py.get(graph, "nodes", []))];
  var edges: any = [...py.iter(py.get(graph, "edges", []))];
  var ontology: any = py.get(semantic_ir, "ontology", {});
  if (py.truthy(py.get(ontology, "primary_domain"))) {
    py.listAppend(nodes, {"id": `domain:${py.toStr(py.at(ontology, "primary_domain"))}`, "type": "domain"});
  }
  if (!py.truthy(nodes)) {
    nodes = [{"id": "semantic:root", "type": "semantic"}];
  }
  return {"ir": "semantic_runtime_graph", "nodes": py.sorted(nodes, {key: ((item: any) => py.toStr(py.get(item, "id", ""))) as (item: any) => any}), "edges": edges, "bounded": true};
}
